zephyr/lib/regions_mm.c

```diff
@@ -1,52 +1,759 @@
+		block_size =
+		    1 << heap->physical_blocks_allocators[mem_block_iterator]->blk_sz_shift;
+		block_count = SOF_DIV_ROUND_UP((uint64_t)alloc_size, (uint64_t)block_size);
+
+		if (block_count >
+		    heap->physical_blocks_allocators[mem_block_iterator]->num_blocks)
+			continue;
+		/* Try span alloc on first available mem_block for non span
+		 * check if block size is sufficient.
+		 */
+		if (heap->allocating_continuously) {
+			allocation_error_code = sys_mem_blocks_alloc_contiguous(
+				heap->physical_blocks_allocators[mem_block_iterator], block_count,
+				&ptr);
+		} else if (block_size > alloc_size) {
+			allocation_error_code = sys_mem_blocks_alloc(
+				heap->physical_blocks_allocators[mem_block_iterator], block_count,
+				&ptr);
+		}
+		/* Save allocation data for purpose of freeing correctly
+		 * afterwards should be changed to a implementation in mem_blocks
+		 * that keeps alloc data there.
+		 */
+		if (allocation_error_code == 0) {
+			/* Mechanism of saving allocation size is a bit complicated and genius
+			 * thought up by Adrian Warecki.
+			 * This explanation is long and will be moved to documentation afterwards.
+			 * Initial setup for mechanism:
+			 * First bit array that saves allocated blocks
+			 * in our system this will be bit array saved in mem_blocks API.
+			 * Second bit array which is of exactly same size of first one
+			 * that will save connection of allocation to size of allocation
+			 * by marking if the bit of allocation is connected to next bit.
+			 *
+			 * Examples: lets have simple 16 bit representation of blocks
+			 * First: 0000000000000000 allocate first four blocks 1111000000000000
+			 * Second:0000000000000000                            1110000000000000
+			 * The allocator saves whether block was allocated and second bit array
+			 * saves information by marking in position of a bit allocated 1 if
+			 * it is a start of the allocation and has more than one
+			 * block allocated, 1 if following bit is also part of allocation.
+			 * 0 for individually allocated block and 0 for block ending allocation
+			 * Continuing the example above
+			 * First: 1111100000000000 allocate another block 1111100000000000
+			 * Second:1110000000000000 after the first one	  1110000000000000
+			 * based on information above we know exactly how long allocations were
+			 * by checking second bit array against first one
+			 * First: 1111111100000000 allocate another block 1111111100000000
+			 * Second:1110000000000000 and another two blocks 1110001000000000
+			 * We are still able to pinpoint allocations size
+			 */
+
+			/* Set bits count - 1 on the same offset as in allocation bit array */
+			allocation_bitarray_offset = (uintptr_t)ptr
+					- (uintptr_t)heap->physical_blocks_allocators
+						[mem_block_iterator]->buffer;
+			allocation_bitarray_position = allocation_bitarray_offset / block_size;
+
+			/* Need to set bits only if we have more than 1 block to allocate */
+			if (block_count - 1)
```

Comment:
That would break the mechanism ? Lets confine the discussion of the allocation size saving mechanism to one comment please. Preferably the first one.

Reply:
using
```
if (block_count > 1)
```
instead of
```
if (block_count - 1)
```
wouldn't break the mechanism. You even wrote in your comment "more than 1 block to allocate" so a more obvious and logical way would be to check exactly that: "more than 1"